frontend/src/metabase/admin/performance/components/StrategyEditorForDatabases.tsx

```diff
@@ -0,0 +1,328 @@
+
+    if (savedStrategy?.type === "duration") {
+      savedStrategy.unit = "hours";
+    }
+
+    const [nextLocation, setNextLocation] = useState<Location>();
+
+    const { show: askConfirmation, modalContent: confirmationModal } =
+      useConfirmation();
+
+    const askBeforeDiscardingChanges = useCallback(
+      (onConfirm: () => void) =>
+        askConfirmation({
+          title: t`Discard your changes?`,
+          message: t`Your changes haven’t been saved, so you’ll lose them if you navigate away.`,
+          confirmButtonText: "Discard",
+          onConfirm,
+        }),
+      [askConfirmation],
+    );
+
+    const [isStrategyFormDirty, setIsStrategyFormDirty] = useState(false);
+    const showStrategyForm = targetId !== null;
+
+    const [isConfirmed, setIsConfirmed] = useState(false);
+
+    useEffect(() => {
+      if (!route) {
+        return;
+      }
+      const removeLeaveHook = router.setRouteLeaveHook(route, location => {
+        if (isStrategyFormDirty && !isConfirmed) {
+          askBeforeDiscardingChanges(() => {
+            setIsConfirmed(true);
+            setNextLocation(location);
+          });
+          return false;
+        }
+      });
+      return removeLeaveHook;
+    }, [
+      router,
+      route,
+      isConfirmed,
+      isStrategyFormDirty,
+      askBeforeDiscardingChanges,
+    ]);
+
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+      if (nextLocation) {
+        dispatch(push(nextLocation));
+      }
+    }, [dispatch, nextLocation]);
+
+    /** Update the targetId (the id of the currently edited model) but confirm if the form is unsaved */
+    const safelyUpdateTargetId: SafelyUpdateTargetId = (
+      newTargetId,
+      isFormDirty,
+    ) => {
+      if (targetId === newTargetId) {
+        return;
+      }
+      const update = () => setTargetId(newTargetId);
+      if (isFormDirty) {
+        askBeforeDiscardingChanges(update);
+      } else {
+        update();
+      }
+    };
+
+    useEffect(() => {
+      if (!canOverrideRootStrategy && targetId === null) {
+        setTargetId(rootId);
+      }
+    }, [canOverrideRootStrategy, targetId]);
+
+    const saveStrategy = useCallback(
+      async (values: Strategy) => {
+        if (targetId === null) {
+          return;
+        }
+
+        const baseConfig: Pick<Config, "model" | "model_id"> = {
+          model: targetId === rootId ? "root" : "database",
+          model_id: targetId,
+        };
+
+        const otherConfigs = configs.filter(
+          config => config.model_id !== targetId,
+        );
+        if (values.type === "inherit") {
+          await CacheConfigApi.delete(baseConfig, { hasBody: true }).then(
+            () => {
+              setConfigs(otherConfigs);
+            },
+          );
+        } else {
+          const validFields = getFieldsForStrategyType(values.type);
+          const newStrategy = pick(values, validFields) as Strategy;
```

Comment:
Why is this needed?

Reply:
Good question. I've added a comment to explain